### PR TITLE
OBSOLETE: Update accessibility modifiers

### PR DESCRIPTION
**Pull request by xtqqczze**
# PR Summary

* Autofix [RCS1102](https://github.com/JosefPihrt/Roslynator/blob/master/docs/analyzers/RCS1102.md): Make class static when only containing static members.
  * plus 4 manual fixes
* Autofix [RCS1225](https://github.com/JosefPihrt/Roslynator/blob/master/docs/analyzers/RCS1225.md): Make class sealed when only having a private constructor.
  * Fix [CA1047](https://docs.microsoft.com/visualstudio/code-quality/ca1047): Do not declare protected members in sealed types
* Change sealed classes with private constructors to static classes
* Revert changes to public APIs causing test failures

## PR Context
https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md#member-conventions

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [NA] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [NA] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [NA] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [NA] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [NA] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by vexx32** [comment]
@xtqqczze you may want to update the PR description with your reasoning for each change, since most of them seem to be on a per-class basis. I'd also recommend a more general PR title, since it seems you're not only adding `static` modifiers here, so it's probably a little less confusing to use something like `Update class modifiers` and explain static / sealed / etc for each class you're changing in the PR description. 🙂

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs [hunk 1]
```diff
@@ -565,7 +565,7 @@ private static string GetNativeDllSubFolderName(out string ext)
     /// <summary>
     /// This is the managed entry point for Microsoft.PowerShell.CoreCLR.AssemblyLoadContext.dll.
     /// </summary>
-    public class PowerShellAssemblyLoadContextInitializer
+    public static class PowerShellAssemblyLoadContextInitializer
```
Comment (iSazonov) [inline src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs]: Breaking change in public API?

Comment (xtqqczze) [reply src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs]: Yes, this change removes the public constructor for the class so is technically a breaking change. Does creating an instance of `PowerShellAssemblyLoadContextInitializer` have any obvious utility?

Comment (iSazonov) [reply src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs]: We need a confirmation from @daxian-dbw

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs [hunk 2]
```diff
@@ -24,7 +24,7 @@ namespace Microsoft.PowerShell.Commands.ShowCommandInternal
     /// Contains information about a cmdlet's Shard ParameterSet,
     /// ParameterSets, Parameters, Common Parameters and error message.
     /// </summary>
-    public class CommandViewModel : INotifyPropertyChanged
+    public sealed class CommandViewModel : INotifyPropertyChanged
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs]: Another public API. @SteveL-MSFT can we change this?

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 3]
```diff
@@ -1774,7 +1774,7 @@ internal static string CheckSuspiciousContent(Ast scriptBlockAst)
             return null;
         }
 
-        class SuspiciousContentChecker
+        static class SuspiciousContentChecker
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: @PaulHigin @TravisEz13 Can we make the class static?

File: src/System.Management.Automation/engine/NullString.cs [hunk 4]
```diff
@@ -6,7 +6,7 @@ namespace System.Management.Automation.Language
     /// <summary>
     /// This type is introduced to provide a way to pass null into a .NET method that has a string parameter.
     /// </summary>
-    public class NullString
+    public sealed class NullString
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/NullString.cs]: Causes test failure due to public API change:

https://dev.azure.com/powershell/PowerShell/_build/results?buildId=45285&view=ms.vss-test-web.build-test-results-tab&runId=1310450&resultId=100810&paneView=debug

File: src/System.Management.Automation/engine/InvocationInfo.cs [hunk 5]
```diff
@@ -451,7 +451,7 @@ internal void ToPSObjectForRemoting(PSObject psObject)
     /// <summary>
     /// A CommandInfo that has been serialized/deserialized as part of an InvocationInfo during a remote invocation.
     /// </summary>
-    public class RemoteCommandInfo : CommandInfo
+    public sealed class RemoteCommandInfo : CommandInfo
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/InvocationInfo.cs]: Causes test failure due to public API change:

https://dev.azure.com/powershell/PowerShell/_build/results?buildId=45285&view=ms.vss-test-web.build-test-results-tab&runId=1310450&resultId=100640&paneView=debug

File: src/System.Management.Automation/engine/MshObject.cs [hunk 6]
```diff
@@ -2457,7 +2457,7 @@ internal enum WriteStreamType : byte
     /// Serves as a placeholder BaseObject when PSObject's
     /// constructor with no parameters is used.
     /// </summary>
-    public class PSCustomObject
+    public sealed class PSCustomObject
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/MshObject.cs]: Causes test failure due to public API change:

https://dev.azure.com/powershell/PowerShell/_build/results?buildId=45285&view=ms.vss-test-web.build-test-results-tab&runId=1310450&resultId=100569&paneView=debug

File: src/System.Management.Automation/engine/ProxyCommand.cs [hunk 7]
```diff
@@ -9,19 +9,8 @@ namespace System.Management.Automation
     /// <summary>
     /// A ProxyCommand class used to represent a Command constructed Dynamically.
     /// </summary>
-    public sealed class ProxyCommand
+    public static class ProxyCommand
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/ProxyCommand.cs]: Public API change. Local test failure: "IsAbstract value is not correct" but no failure on CI!

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs (hunk 2 above)
Comment (xtqqczze) [reply src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs]: Namespace is `Microsoft.PowerShell.Commands.ShowCommandInternal` so change is Bucket 4: "Clearly Non-Public"?

File: src/System.Management.Automation/engine/NullString.cs (hunk 4 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/NullString.cs]: It is in SDK. It can be a huge breaking change. Why do we really need the PS's changes?

**Review by iSazonov — COMMENTED** [review]
Why do we set `sealed` for internal classes? Can you add more info in the PR description?

File: src/System.Management.Automation/engine/NullString.cs (hunk 4 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/NullString.cs]: reverted

File: src/System.Management.Automation/engine/InvocationInfo.cs (hunk 5 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/InvocationInfo.cs]: reverted

File: src/System.Management.Automation/engine/MshObject.cs (hunk 6 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/MshObject.cs]: reverted

File: src/System.Management.Automation/engine/ProxyCommand.cs (hunk 7 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/ProxyCommand.cs]: reverted

File: src/System.Management.Automation/engine/NullString.cs (hunk 4 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/NullString.cs]: @iSazonov Maybe a change to the API would not be so bad. In fact subclassing `NullString` would be a bad idea so maybe we should disallow this?

Comment (iSazonov) [reply src/System.Management.Automation/engine/NullString.cs]: Maybe. It is better to open new issue and discuss there (and get PowerShell Committee conclusion)

**Comment by iSazonov** [comment]
@xtqqczze Have you plans to continue? Please resolve merge conflicts.

**Comment by xtqqczze** [comment]
@iSazonov I have seperated RCS1018 changes into #12820 and will continue this later.

**Comment by iSazonov** [comment]
Stil lWIP?

**Comment by iSazonov** [comment]
@xtqqczze Please rebase.

**Comment by xtqqczze** [comment]
@iSazonov rebased

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs (hunk 2 above)
Comment (iSazonov) [reply src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs:27]: The code is frozen. Please revert the change.

File: src/Microsoft.Management.UI.Internal/commandHelpers/ShowCommandHelper.cs [hunk 8]
```diff
@@ -23,7 +23,7 @@ namespace Microsoft.PowerShell.Commands.ShowCommandInternal
     /// <summary>
     /// Implements thw WPF window part of the show-command cmdlet.
     /// </summary>
-    internal class ShowCommandHelper : IDisposable
+    internal sealed class ShowCommandHelper : IDisposable
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/commandHelpers/ShowCommandHelper.cs:26]: The same.

**Comment by iSazonov** [comment]
@daxian-dbw Could you please review?

**Comment by xtqqczze** [comment]
@daxian-dbw Can you please review?

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs (hunk 2 above)
Comment (xtqqczze) [reply src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs:27]: When was `Microsoft.Management.UI.Internal` frozen?

Comment (iSazonov) [reply src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs:27]: The code was added only for backward compatibility. All new work will be in PowerShell/GraphicalTools repo.

**Comment by daxian-dbw** [comment]
@PowerShell/powershell-maintainers reviewed this and agreed to accept this PR as long as a thorough review is done. Please delay merging it after the 7.1.0-RC.1 release branch is created.

**Comment by iSazonov** [comment]
Convert to draft until 7.1 rc1.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 3 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: The class contains only static members.

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs (hunk 1 above)
Comment (xtqqczze) [reply src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs:569]: reverted

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 3 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: @iSazonov Should I revert?

**Comment by xtqqczze** [comment]
rebased

**Comment by xtqqczze** [comment]
CodeFactor "complex method" and "complex code" issues are not new.

**Review by iSazonov — APPROVED** [review]

**Review by TravisEz13 — COMMENTED** [review]
Too many different types of changes in one PR.  This greatly slows down review.

**Comment by TravisEz13** [comment]
Please separate the static and sealed changes.

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs (hunk 1 above)
Comment (xtqqczze) [reply src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs]: Opened #13874

**Comment by xtqqczze** [comment]
I will separate changes into new PRs, converting to draft.

**Comment by iSazonov** [comment]
> I will separate changes into new PRs, converting to draft.

Have you any updates?

**Comment by xtqqczze** [comment]
> > I will separate changes into new PRs, converting to draft.
> 
> Have you any updates?

* WIP: #13897
* Merged: #13874

**Comment by iSazonov** [comment]
> > > I will separate changes into new PRs, converting to draft.
> > 
> > 
> > Have you any updates?
> 
> * WIP: #13897
> * Merged: #13874

So the PR can be closed?

**Comment by iSazonov** [comment]
> > > > I will separate changes into new PRs, converting to draft.
> > > 
> > > 
> > > Have you any updates?
> > 
> > 
> > 
> > * WIP: #13897
> > * Merged: #13874
> 
> So the PR can be closed?

@xtqqczze Can we close?

**Comment by xtqqczze** [comment]
> @xtqqczze Can we close?

This PR has some changes that depends on #13897, so it should stay as draft.

**Comment by xtqqczze** [comment]
RCS1102 changes are split to #14335.

**Comment by xtqqczze** [comment]
e0e60ba88a05045f486af49615916f48bd396a1d changes submitted as #15671.

**Comment by xtqqczze** [comment]
2069d92b3e2760ab631efbf12bbde9ca46e661e2 changes submitted as #15675

**Comment by xtqqczze** [comment]
Changes from this PR are now either merged or submitted in PRs.

**Pull request closed without merging**